src/measure_theory/function/intersectivity.lean

```diff
@@ -0,0 +1,122 @@
+  let N : set α := ⋃ u : finset ℕ, M (set.indicator (⋂ n ∈ u, s n) 1),
+  have hN₀ : μ N = 0 := measure_Union_null (λ u, meas_snorm_ess_sup_lt),
+  have hN₁ : ∀ u : finset ℕ, ((⋂ n ∈ u, s n) \ N).nonempty → 0 < μ (⋂ n ∈ u, s n),
+  { simp_rw pos_iff_ne_zero,
+    rintro u ⟨x, hx⟩ hu,
+    refine hx.2 (mem_Union.2 ⟨u, (_ : _ < _)⟩),
+    rw [indicator_of_mem hx.1, snorm_ess_sup_eq_zero_iff.2],
+    simp,
+    { rwa [indicator_ae_eq_zero, function.support_one, inter_univ] } },
+  let f : ℕ → α → ℝ≥0∞ := λ n, (↑(n + 1) : ℝ≥0∞)⁻¹ • ∑ k in finset.range (n + 1), (s k).indicator 1,
+  have hfapply : ∀ n a, f n a = (↑(n + 1))⁻¹ * ∑ k in finset.range (n + 1), (s k).indicator 1 a,
+  { simp only [f, pi.coe_nat, pi.smul_apply, pi.inv_apply, finset.sum_apply, eq_self_iff_true,
+    forall_const, implies_true_iff, smul_eq_mul] },
+  have hf₀ : 0 ≤ f,
+  { exact zero_le _ },
+  have hf' : ∀ n, measurable (∑ k in finset.range n, (s k).indicator 1 : α → ℝ≥0∞),
+  { exact λ n, (finset.measurable_sum' _ $ λ i _, measurable_one.indicator $ hs i) },
+  have hf : ∀ n, measurable (f n),
+  { refine λ n, measurable.mul' _
+      (finset.measurable_sum' _ $ λ i _, measurable_one.indicator $ hs i),
+    exact @measurable_const ℝ≥0∞ _ _ _ (↑(n + 1))⁻¹ },
+  have hf₁ : ∀ n, f n ≤ 1,
+  { rintro n a,
+    rw [hfapply, ←ennreal.div_eq_inv_mul],
+    refine (ennreal.div_le_iff_le_mul (or.inl $ nat.cast_ne_zero.2 n.succ_ne_zero) $
+      or.inr one_ne_zero).2 _,
+    rw [mul_comm, ←nsmul_eq_mul, ←finset.card_range n.succ],
+    exact finset.sum_le_card_nsmul _ _ _ (λ _ _, indicator_le (λ _ _, le_rfl) _) },
+  have hf₂ : ∀ n, r ≤ ∫⁻ a, f n a ∂μ,
+  { rintro n,
+    simp_rw hfapply,
+    rw [lintegral_const_mul, lintegral_finset_sum],
+    simp only [lintegral_indicator_one (hs _)],
+    rw ←ennreal.div_eq_inv_mul,
+    refine (ennreal.le_div_iff_mul_le (or.inl $ nat.cast_ne_zero.2 n.succ_ne_zero) $ or.inl $
+      ennreal.nat_ne_top _).2 _,
+    refine le_trans _ (finset.card_nsmul_le_sum _ _ _ $ λ _ _, hr _),
+    { simp [mul_comm] },
+    { exact λ _ _, measurable_one.indicator (hs _) },
+    { exact finset.measurable_sum _ (λ _ _, measurable_one.indicator $ hs _) } },
+  have hμ : μ ≠ 0,
+  { unfreezingI { rintro rfl },
+    exact hr₀ (le_bot_iff.1 $ hr 0) },
+  obtain ⟨x, hxN, hx⟩ := exists_not_mem_null_laverage_le hμ
+    (ne_top_of_le_ne_top (measure_ne_top μ univ) _) hN₀,
+  rw laverage_eq at hx,
+  replace hx := (ennreal.div_le_div_right ((le_limsup_of_le ⟨μ univ, eventually_map.2 _⟩ $ λ b hb,
+    _).trans $ limsup_lintegral_le hf (λ n, ae_of_all μ $ hf₁ n) $
+    ne_of_eq_of_ne lintegral_one is_finite_measure.measure_univ_lt_top.ne) _).trans hx,
+  refine ⟨{n | x ∈ s n}, λ hxs, _, λ u hux hu, _⟩,
+  -- This next block proves that a set of strictly positive natural density is infinite, mixed with
+  -- the fact that `{n | x ∈ s n}` has strictly positive natural density.
+  -- TODO: Separate it out to a lemma once we have a natural density API.
```

Comment:
IMHO, you should separate it to a lemma now, then migrate to a new API when it's available.

Reply:
Eh I'm not sure how to go about this. There's some plumbing involved and figuring it out is probably as hard as writing the density API in the first place. The density API is on my todo list, so I'm definitely not forgetting that todo.